include/devicetree: clocks: Add DT_CLOCKS_HAS_FOO & DT_NUM_CLOCKS

erwango opened the pull request:
include/devicetree: clocks: Add DT_CLOCKS_HAS_FOO & DT_NUM_CLOCKS
    
Add macros DT_CLOCKS_HAS_NAME and DT_CLOCKS_HAS_IDX.
These macros allow to check the presence of a specific clock name or a specific cell at a given index in a clocks property.
Matching _INST_ macro variants are also provided.
    
Also add DT_NUM_CLOCKS and its _INST_ variant that counts the number of clocks available for a given node.

Add matching tests.

erwango committed:
include/devicetree: clocks: Add DT_CLOCKS_HAS_FOO & DT_NUM_CLOCKS

Add macros DT_CLOCKS_HAS_NAME and DT_CLOCKS_HAS_IDX.
These macros allow to check the presence of a specific clock name
or a specific cell at a given index in a clocks property.
Matching _INST_ macro variants are also provided.

Also add DT_NUM_CLOCKS and its _INST_ variant that counts
the number of clocks available for a given node.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/lib/devicetree: api: Add tests for new DT_{INST_}CLOCKS macros

Add matching tests for newly added DT_{INST_}CLOCKS macros.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -22,6 +24,92 @@ extern "C" {
+ * @return 1 if the clock name exists, 0 otherwise
+ */
+#define DT_CLOCKS_HAS_NAME(node_id, name) \
+	IS_ENABLED(DT_CAT4(node_id, _P_clocks_NAME_, name, _EXISTS))
```
Why not add a general-purpose helper for deciding if a named property exists, like `DT_PROP_HAS_NAME(node_id, prop, name)`, and using it here?

erwango replied:
Sure, will add `DT_PROP_HAS_NAME`

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -246,14 +334,39 @@ extern "C" {
  */
 #define DT_CLOCKS_CELL(node_id, cell) DT_CLOCKS_CELL_BY_IDX(node_id, 0, cell)
 
+/**
+ * @brief Equivalent to DT_CLOCKS_HAS_IDX(node_id, idx)
+ * @param inst DT_DRV_COMPAT instance number; may or may not have any clocks property
+ * @param idx index of a clocks property phandle-array whose existence to check
+ * @return 1 if the index exists, 0 otherwise
+ */
+#define DT_INST_CLOCKS_HAS_IDX(inst, idx) \
+	DT_CLOCKS_HAS_IDX(DT_DRV_INST(inst), idx)
+
+/**
+ * @brief Equivalent to DT_CLOCK_HAS_NAME(node_id, name)
+ * @param inst DT_DRV_COMPAT instance number; may or may not have any clock-names property.
+ * @param name lowercase-and-underscores clock-names cell value name to check
+ * @return 1 if the clock name exists, 0 otherwise
+ */
+#define DT_INST_CLOCKS_HAS_NAME(inst, name) \
+	DT_CLOCKS_HAS_NAME(DT_DRV_INST(inst), name)
+
+/**
+ * @brief Equivalent to DT_NUM_CLOCKS(node_id)
+ * @param inst instance number
+ * @return number of elements in the clocks property
+ */
+#define DT_INST_NUM_CLOCKS(inst) \
+	DT_NUM_CLOCKS(DT_DRV_INST(inst))
+
 /**
  * @brief Get the node identifier for the controller phandle from a
  *        "clocks" phandle-array property at an index
  *
  * @param inst instance number
  * @param idx logical index into "clocks"
- * @return the node identifier for the clock controller referenced at
- *         index "idx"
+ * @return the node identifier for the clock controller referenced at index "idx"
```
Please don't mix whitespace and functional changes in a single commit.

mbolivar-nordic commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1564,6 +1564,17 @@ static void test_clocks(void)
 	zassert_true(DT_SAME_NODE(DT_CLOCKS_CTLR_BY_NAME(TEST_TEMP, clk_b),
 				  DT_NODELABEL(test_clk)), "");
 
+	/* DT_NUM_CLOCKS */
+	zassert_equal(DT_NUM_CLOCKS(TEST_TEMP), 3, "");
```
I'd like to see coverage where the number of clocks is 0 added here and for the INST version, please.

erwango replied:
For now, it is not possible to call this macro on a node with no "clocks" property.
I could possibly implement `DT_NUM_CLOCKS` using `DT_PROP_LEN_OR` which would return 0 is no clocks are provided, but I'm not sure this is really valid (empty clocks property vs no clocks property).
I could also mention this in macro docstring but this is likely valid for a lot of macros on other properties which don't have a fallback. So I'm not really sure this is required.

So is that really required ?

mbolivar-nordic replied:
> So is that really required ?

No, that's fine, let's ignore this request from me then. Thanks for justification

mbolivar-nordic reviewed (changes requested):
First pass review. No big objections since this is just adding new clock-specific convenience helpers; just some initial comments about implementation style. Will look at the docstrings in a follow-up review.

erwango committed:
include/devicetree.h: Add DT_PROP_HAS_NAME macro

Add a DT_PROP_HAS_NAME which helps verifying if a given string
matches an existing value in a prop-names type property.

Also add matching DT_INST_PROP_HAS_NAME

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on include/devicetree/clocks.h:
```diff
@@ -22,6 +24,92 @@ extern "C" {
+ * @return number of elements in the property
+ */
+#define DT_NUM_CLOCKS(node_id) \
+	DT_CAT3(node_id, _P_clocks, _LEN)
```
Yep, will fix

erwango committed:
tests/lib/devicetree: api: Add tests for DT_{INST_}PROP_HAS_NAME

Add tests for DT_PROP_HAS_NAME and DT_INST_PROP_HAS_NAME
macros

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/deviceree: dma: Update using DT_PROP_HAS_NAME

Following introduction of DT_PROP_HAS_NAME,
update DT_DMAS_HAS_NAME definition to reuse
this helper.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -608,6 +608,24 @@
 	IS_ENABLED(DT_CAT6(node_id, _P_, prop, _IDX_, idx, _EXISTS))
 
+/**
+ * @brief Is name "name" available in a foo-names property?
```
Can you give an example devicetree fragment and usage, like we do for other macros? This description is a bit abstract and I think an example would make it easier to use.

erwango replied:
Example added

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -608,6 +608,24 @@
+ *
+ * - interrupts property: use DT_IRQ_HAS_NAME(node_id, idx) instead
+ *
+ * It is an error to use this macro with the interrupts properties.
```
s/properties/property/

erwango replied:
Thanks

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -166,6 +254,32 @@ extern "C" {
+	DT_CLOCKS_HAS_NAME(DT_DRV_INST(inst), name)
+
+/**
+ * @brief Equivalent to DT_NUM_CLOCKS(node_id)
```
s/node_id/DT_DRV_INST(inst)/, here and above

erwango replied:
Ok, sure. Done

mbolivar-nordic commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -737,6 +744,13 @@ static void test_phandles(void)
+	zassert_true(DT_INST_PROP_HAS_NAME(TEST_PH, foos, a), "");
+	zassert_false(DT_INST_PROP_HAS_NAME(TEST_PH, foos, b-c), "");
+	zassert_true(DT_INST_PROP_HAS_NAME(TEST_PH, foos, b_c), "");
+	zassert_false(DT_INST_PROP_HAS_NAME(TEST_PH, bazs, jane), "");
```
Looks like the fix for this got squashed into the wrong commit. It's fixed later on in ` tests/lib/devicetree: api: Add tests for new DT_{INST_}CLOCKS macros `, but it shouldn't be.

erwango replied:
Oh, thanks. Fixed

mbolivar-nordic reviewed (changes requested):
One last round of review; almost there from my perspective. Thanks again.

erwango commented:
Thanks @mbolivar-nordic for the review. I hope everything is fine now.

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -12,6 +12,8 @@
 #ifndef ZEPHYR_INCLUDE_DEVICETREE_CLOCKS_H_
 #define ZEPHYR_INCLUDE_DEVICETREE_CLOCKS_H_
 
+#include <sys/util_macro.h>
+
```
Sorry that I missed this last time, but can you explain why you are adding this include here? It doesn't seem to be needed.

erwango replied:
Good point, IIRC it was required initially while I was not based in devicetree.h macros.
Now this could be removed. Done

erwango commented:
@gmarull would you mind re-approving ?
